Docs/Developer Guide.md

```diff
@@ -0,0 +1,47 @@
+## Overview
+
+## Producer/Consumer Design Pattern
+
+The **Routing and Faulting Custom Device** and **SLSC Switch Custom Device** use a [producer/consumer design pattern](http://www.ni.com/tutorial/3023/en/). The **Routing and Faulting Custom Device** is the producer sending messages to multiple **SLSC Switch Custom Devices** acting as the consumers. The producer monitors changes to VeriStand routing channels and sends disconnect/connect messages to the corresponding consumers. The consumers receive the messages and call into underlying hardware to disconnect/connect endpoints. Existing endpoints are disconnected before new endpoints are connected. Connections that remain unchanged do not glitch.
+
+![Switch Messages](Switch%20Messages.png)
+
+The producer/consumer design pattern is implemented in the [Custom Device Message Library](https://github.com/ni/niveristand-custom-device-message-library). The **Custom Device Message Library** uses named message queues to send requests and receive responses. The **Custom Device Message Library** defines base classes for the producer, consumer, and messages. It also defines derived classes for switch consumer and switch messages. The [SLSC Switch Message Library](https://github.com/ni/niveristand-slsc-switch-message-library) overrides the switch consumer class and calls into the **SLSC Switch API** to connect and disconnect endpoints.
+
+![Class Diagram](Class%20Diagram.png)
+
+## Adding New Hardware Support
+
+The **Routing and Faulting Custom Device** supports **SLSC Switch** hardware. The **Routing and Faulting Custom Device** is designed to support additional hardware with minimal effort. Override the abstract switch consumer class and implement the following hardware specific methods:
+
+| Method Name | Description |
+|---|---|
+| Initialize with Switch Parameters Core | Open a hardware session using the switch parameters. The switch parameters include the resource name and topology. |
+| Connect | Connect the specified endpoints. |
+| Disconnect | Disconnect the specified endpoints. |
+| Disconnect All | Disconnect all connected endpoints. |
+| Shutdown Core | Reset the hardware state and close the hardware session. |
+
+Create a hardware specific custom device based on the asynchronous template with the following properties in the system explorer:
```

Comment:
Nothing dictates that a new custom device has to be asynchronous. It could be inlined, it just might mess with your PCL rate. Or you could have something like Inline Async.

Reply:
The wording is ambiguous, but I wanted to suggest not putting this inline. Maybe it should read: 

Create a hardware specific custom device based on the inline async or asynchronous template... Using the inline template may adversely affect loop rates.

Reply:
I'm fine with that.